### PR TITLE
CI: add actuated-arm64

### PR DESCRIPTION
**Pull request by AkihiroSuda**
See <https://actuated.dev/blog/arm-ci-cncf-ampere>. 
Thanks to Alex Ellis, Ampere Computing, and Equinix.

- - -

Host information:
* CPU: aarch64 (ARMv8)
* Kernel: 5.10.201
  * Lacks ~`CONFIG_CHECKPOINT_RESTORE`~, `CONFIG_BLK_CGROUP_IOCOST`, etc.
* Cgroup: v2
* OS: Ubuntu 22.04
  * Lacks `newuidmap`, `newgidmap`, etc. (still apt-gettable)
  * `sshd` is not running

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -23,7 +23,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        os: [ubuntu-20.04, ubuntu-22.04]
+        os: [ubuntu-20.04, ubuntu-22.04, actuated-arm64-4cpu-12gb]
```
Comment (AkihiroSuda) [inline .github/workflows/test.yml:26]: GitHub App has to be installed from https://github.com/apps/self-actuated

**Comment by AkihiroSuda** [comment]
cc @alexellis @vielmetti

File: .github/workflows/test.yml (hunk 1 above)
Comment (AkihiroSuda) [reply .github/workflows/test.yml:26]: @caniszczyk @cyphar May I install this?

Comment (cyphar) [reply .github/workflows/test.yml:26]: @amye also has the necessary permissions IIRC. I can't install it unfortunately, but I've sent a request.

**Comment by alexellis** [comment]
Thanks for the ping. The build should run once the GitHub app is installed.

File: .github/workflows/test.yml (hunk 1 above)
Comment (amye) [reply .github/workflows/test.yml:26]: Try it now, but "This GitHub App is for customers with an active Actuated subscription."

**Comment by alexellis** [comment]
Thanks to @amye the app is now installed.

File: .github/workflows/test.yml (hunk 1 above)
Comment (AkihiroSuda) [reply .github/workflows/test.yml:26]: Thanks, the app seems now installed

**Comment by alexellis** [comment]
@AkihiroSuda have you had a chance to try out the telemetry to see how much RAM/vCPU is a good fit?

It also works on hosted runners - https://gist.github.com/alexellis/1f33e581c75e11e161fe613c46180771

**Comment by AkihiroSuda** [comment]
> @AkihiroSuda have you had a chance to try out the telemetry to see how much RAM/vCPU is a good fit?
> 
> It also works on hosted runners - https://gist.github.com/alexellis/1f33e581c75e11e161fe613c46180771

### Rootful
https://github.com/opencontainers/runc/actions/runs/7223805271/job/19683690706?pr=4142
```
Post job cleanup.
Total RAM: 11.38GB
Total vCPU: 4

Load averages:
Max 1 min: 4.71 (117.75%)
Max 5 min: 1.57 (39.25%)
Max 15 min: 0.75 (18.75%)

RAM usage (70 samples):
Max RAM usage: 675.4MB

Max 10s avg RAM usage: 588.3MB
Max 1 min avg RAM usage: 487.5MB
Max 5 min avg RAM usage: 456.2MB

Disk read: 410.1MB
Disk write: 1.528GB
Max disk I/O inflight: 0
Free: 26.27GB	Used: 3.626GB	(Total: 31.53GB)	

Egress adapter RX: 183MB
Egress adapter TX: 1.896MB

Entropy min: 256
Entropy max: 256

Max open connections: 14
Max open files: 1088
Processes since boot: 72968

Run time: 5m46s
```

### Rootless
https://github.com/opencontainers/runc/actions/runs/7223805271/job/19683690402?pr=4142
```
Post job cleanup.
Total RAM: 11.38GB
Total vCPU: 4

Load averages:
Max 1 min: 3.70 (92.50%)
Max 5 min: 1.13 (28.25%)
Max 15 min: 0.68 (17.00%)

RAM usage (115 samples):
Max RAM usage: 589.7MB

Max 10s avg RAM usage: 530.6MB
Max 1 min avg RAM usage: 468.9MB
Max 5 min avg RAM usage: 447.4MB

Disk read: 410.3MB
Disk write: 1.418GB
Max disk I/O inflight: 0
Free: 26.33GB	Used: 3.568GB	(Total: 31.53GB)	

Egress adapter RX: 184.7MB
Egress adapter TX: 1.531MB

Entropy min: 256
Entropy max: 256

Max open connections: 22
Max open files: 1280
Processes since boot: 121033

Run time: 9m34s
```

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -53,6 +53,22 @@ jobs:
     runs-on: ${{ matrix.os }}
 
     steps:
+# https://gist.github.com/alexellis/1f33e581c75e11e161fe613c46180771#file-metering-gha-md
+# vmmeter start
+    - uses: alexellis/arkade-get@master
+      if: matrix.os == 'actuated-arm64-4cpu-12gb'
```
Comment (alexellis) [inline .github/workflows/test.yml:59]: If you're interested - the metering also works on hosted runners.

**Comment by alexellis** [comment]
It looks like at least the 4vCPU are being used in the build, so you could try upping to 6vCPU or 8vCPU.

RAM usage is low, so that limit can probably stay, or get dipped down to 8gb.

File: .github/workflows/test.yml (hunk 2 above)
Comment (AkihiroSuda) [reply .github/workflows/test.yml:59]: Thanks, but probably we do not need this for hosted runners, as we can't change the instance type for hosted runners anyway

**Comment by AkihiroSuda** [comment]
> It looks like at least the 4vCPU are being used in the build, so you could try upping to 6vCPU or 8vCPU.
> 
> RAM usage is low, so that limit can probably stay, or get dipped down to 8gb.

How to do this?
Are there other actuated-arm64-%dcpu-%dgb runners?

**Comment by alexellis** [comment]
Hi @AkihiroSuda please see the announcement that you linked to, it explains how to pick the values, additionally the docs do too.

https://docs.actuated.dev/examples/custom-vm-size/

**Comment by AkihiroSuda** [comment]
Thanks, updated to 6cpu-8gb

**Review by kolyshkin — CHANGES_REQUESTED** [review]
Impressive work! Left some comments. 

In particular, I found some more issues with check-config.sh, and also cherry-picked the changes from here (@AkihiroSuda PTAL https://github.com/opencontainers/runc/pull/4152 -- I think we can merge that one first and then rebase this one, less code to review this way).

File: script/check-config.sh [hunk 3]
```diff
@@ -43,6 +48,9 @@ is_set_as_module() {
 }
 
 color() {
+	if [ "$NO_COLOR" = "1" ]; then
+		return
+	fi
```
[the remark this answers is not in the document]
Comment (kolyshkin) [reply script/check-config.sh:53]: Impemented in https://github.com/opencontainers/runc/pull/4152

File: script/check-config.sh [hunk 4]
```diff
@@ -4,6 +4,11 @@ set -e -u
 # bits of this were adapted from check_config.sh in docker
 # see also https://github.com/docker/docker/blob/master/contrib/check-config.sh
 
+: "${NO_COLOR:=}"
+if [ -z "$NO_COLOR" ] && ! [ -t 1 ]; then
+	NO_COLOR=1
+fi
+
```
[the remark this answers is not in the document]
Comment (kolyshkin) [reply script/check-config.sh:11]: Implemented in https://github.com/opencontainers/runc/pull/4152

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -40,19 +41,77 @@ jobs:
...
           - dmz: runc_nodmz
             os: ubuntu-20.04
+          - go-version: 1.20.x
+            os: actuated-arm64-6cpu-8gb
+          - race: "-race"
+            os: actuated-arm64-6cpu-8gb
+          - criu: criu-dev
+            os: actuated-arm64-6cpu-8gb
+          - dmz: runc_nodmz
+            os: actuated-arm64-6cpu-8gb
```
[the remark this answers is not in the document]
Comment (kolyshkin) [reply .github/workflows/test.yml:51]: No need to; I guess I see why.

File: tests/integration/helpers.bash [hunk 6]
```diff
@@ -450,6 +450,22 @@ function requires() {
...
+			local p f f2
+			init_cgroup_paths
+			if [ -v CGROUP_V1 ]; then
+				p="$CGROUP_CPU_BASE_PATH"
+				f="blkio.weight"
+				f2="blkio.bfq.weight"
+			elif [ -v CGROUP_V2 ]; then
+				p="$CGROUP_BASE_PATH"
+				f="io.weight"
+				f="io.bfq.weight"
```
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply tests/integration/helpers.bash]: fixed

File: tests/integration/helpers.bash [hunk 7]
```diff
@@ -450,6 +450,22 @@ function requires() {
 				skip_me=1
 			fi
 			;;
+		cgroups_blk_iocost)
```
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply tests/integration/helpers.bash]: fixed

File: update.go [hunk 8]
```diff
@@ -147,26 +147,27 @@ other options are ignored.
...
-				Reservation:       i64Ptr(0),
-				Swap:              i64Ptr(0),
-				Kernel:            i64Ptr(0),
-				KernelTCP:         i64Ptr(0),
-				CheckBeforeUpdate: boolPtr(false),
+				Limit:             nil,
+				Reservation:       nil,
+				Swap:              nil,
+				Kernel:            nil,
+				KernelTCP:         nil,
```
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply update.go]: fixed

File: update.go [hunk 9]
```diff
@@ -147,26 +147,27 @@ other options are ignored.
...
-				Burst:           u64Ptr(0),
-				Period:          u64Ptr(0),
-				RealtimeRuntime: i64Ptr(0),
-				RealtimePeriod:  u64Ptr(0),
+				Shares:          nil,
+				Quota:           nil,
+				Burst:           nil,
+				Period:          nil,
+				RealtimeRuntime: nil,
+				RealtimePeriod:  nil,
```
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply update.go]: fixed

File: .github/workflows/test.yml [hunk 10]
```diff
@@ -57,13 +57,13 @@ jobs:
...
         sudo apt update
-        sudo apt install libseccomp-dev criu sshfs
+        sudo apt install -y libseccomp-dev criu sshfs
 
     - name: install deps (criu ${{ matrix.criu }})
       if: matrix.criu != ''
       run: |
         sudo apt -q update
-        sudo apt -q install libseccomp-dev sshfs \
+        sudo apt -q install -y libseccomp-dev sshfs \
```
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply .github/workflows/test.yml]: fixed

File: update.go [hunk 11]
```diff
@@ -147,26 +147,27 @@ other options are ignored.
...
 				Cpus:            "",
 				Mems:            "",
 			},
 			BlockIO: &specs.LinuxBlockIO{
-				Weight: u16Ptr(0),
+				Weight: nil,
```
[the remark this answers is not in the document]
Comment (AkihiroSuda) [reply update.go]: fixed

File: tests/integration/testdata/seccomp_syscall_test1.c [hunk 12]
```diff
@@ -57,10 +57,10 @@ int main(void)
...
 
 	// Multiple arguments with AND rules.
-	syscall_assert(raw(kcmp, 0, 1337, 0, 0, 0), -ESRCH);
-	syscall_assert(raw(kcmp, 0, 0, 0, 0, 0), -EPERM);
-	syscall_assert(raw(kcmp, 500, 1337, 0, 0, 0), -EPERM);
-	syscall_assert(raw(kcmp, 500, 500, 0, 0, 0), -EPERM);
+	syscall_assert(raw(ftruncate, 123456789, 1337), -EBADF);
+	syscall_assert(raw(ftruncate, 0, 0), -EPERM);
+	syscall_assert(raw(ftruncate, 500, 1337), -EPERM);
+	syscall_assert(raw(ftruncate, 500, 500), -EPERM);
```
Comment (kolyshkin) [inline tests/integration/testdata/seccomp_syscall_test1.c:63]: I think the original idea here was to check that AND rules work. Meaning, (1) if arg0 is 0 and arg1 is 1337, the syscall is allowed (and we get a different error), while (2) if either arg0 is not 0 or arg1 is not 1337, we should get EPERM, and (3) if both arg0 and arg1 are not as expected, we also get EPERM.

This change removed the check for (2) for the case when arg0 is as expected and arg1 is not.

In other words, the second check for ftruncate should look like this:
```C
syscall_assert(raw(ftruncate, 123456789, 0), -EPERM);
```

(cc @cyphar)

File: update.go [hunk 13]
```diff
@@ -265,19 +248,31 @@ other options are ignored.
...
+			// To avoid error "unable to set swap limit without memory limit"
+			r.Memory.Swap = i64Ptr(0)
+		}
+		if r.CPU.Idle != nil && r.CPU.Shares == nil {
+			// To avoid error "failed to write \"4\": write /sys/fs/cgroup/runc-cgroups-integration-test/test-cgroup-7341/cpu.weight: invalid argument"
+			r.CPU.Shares = u64Ptr(0)
+		}
+
+		if (r.Memory.Kernel != nil && *r.Memory.Kernel != 0) || (r.Memory.KernelTCP != nil && *r.Memory.KernelTCP != 0) {
 			logrus.Warn("Kernel memory settings are ignored and will be removed")
```
Comment (kolyshkin) [inline update.go:269]: I think it's enough to check for `!= nil` here (as we do ignore `--kernel-memory 0`, but were unable to distinguish it from the default value of 0 before -- now we are able to).

File: update.go [hunk 14]
```diff
@@ -265,19 +248,31 @@ other options are ignored.
...
-		if *r.Memory.Kernel != 0 || *r.Memory.KernelTCP != 0 {
+		// Fix up values
+		if r.Memory.Limit != nil && *r.Memory.Limit == -1 && r.Memory.Swap == nil {
+			// To avoid error "unable to set swap limit without memory limit"
+			r.Memory.Swap = i64Ptr(0)
+		}
+		if r.CPU.Idle != nil && r.CPU.Shares == nil {
+			// To avoid error "failed to write \"4\": write /sys/fs/cgroup/runc-cgroups-integration-test/test-cgroup-7341/cpu.weight: invalid argument"
+			r.CPU.Shares = u64Ptr(0)
+		}
```
Comment (kolyshkin) [inline update.go:266]: I'm not sure about this, need to look more closely. Perhaps those errors are actually valid?

I barely remember that runc should reuse the previously set limit (if set).

Will take a closer look later.

**Review by kolyshkin — CHANGES_REQUESTED** [review]
Left some comments. In particular, I am concerned about the change in `runc update: distinguish nil from zero` (which I need to take a deeper look at). Other stuff is minor/nits.

File: tests/integration/testdata/seccomp_syscall_test1.c (hunk 12 above)
Comment (AkihiroSuda) [reply tests/integration/testdata/seccomp_syscall_test1.c]: Fixed

File: update.go (hunk 13 above)
Comment (AkihiroSuda) [reply update.go]: Fixed

**Comment by AkihiroSuda** [comment]
Rebased

File: update.go (hunk 14 above)
Comment (AkihiroSuda) [reply update.go:266]: What should we do?

**Comment by AkihiroSuda** [comment]
Looks like they enabled `CONFIG_CHECKPOINT_RESTORE` recently, but CRIU still does not work
https://github.com/opencontainers/runc/actions/runs/7900308834/job/21561582928?pr=4142

```
ok 25 runc run/create should error for a non-empty cgroup
ok 26 runc run/create should refuse pre-existing frozen cgroup
not ok 27 checkpoint and restore
# (from function `simple_cr' in file tests/integration/checkpoint.bats, line 114,
#  in test file tests/integration/checkpoint.bats, line 130)
#   `simple_cr' failed
# runc spec (status=0):
#
# runc run -d --console-socket /tmp/bats-run-qgKMpl/runc.dwPbfF/tty/sock test_busybox (status=0):
#
# runc state test_busybox (status=0):
# {
#   "ociVersion": "1.1.0+dev",
#   "id": "test_busybox",
#   "pid": 23274,
#   "status": "running",
#   "bundle": "/tmp/bats-run-qgKMpl/runc.dwPbfF/bundle",
#   "rootfs": "/tmp/bats-run-qgKMpl/runc.dwPbfF/bundle/rootfs",
#   "created": "2024-02-14T11:13:42.65358032Z",
#   "owner": ""
# }
# runc checkpoint --work-path ./work-dir test_busybox (status=1):
# time="2024-02-14T11:13:42Z" level=warning msg="--- Quoting \"work-dir/dump.log\""
# time="2024-02-14T11:13:42Z" level=warning msg="252:(00.027091) net: Unlock network"
# time="2024-02-14T11:13:42Z" level=warning msg="253:(00.027097) Running network-unlock scripts"
# time="2024-02-14T11:13:42Z" level=warning msg="254:(00.027101) \tRPC"
# time="2024-02-14T11:13:42Z" level=warning msg="255:(00.032201) Unfreezing tasks into 1"
# time="2024-02-14T11:13:42Z" level=warning msg="256:(00.032262) \tUnseizing 23274 into 1"
# time="2024-02-14T11:13:42Z" level=warning msg="257:(00.032319) Error (criu/cr-dump.c:2098): Dumping FAILED."
# time="2024-02-14T11:13:42Z" level=warning msg=---
# time="2024-02-14T11:13:42Z" level=error msg="criu failed: type DUMP errno 0"
# (00.027091) net: Unlock network
# (00.027097) Running network-unlock scripts
# (00.027101) 	RPC
# (00.032201) Unfreezing tasks into 1
# (00.032262) 	Unseizing 23274 into 1
# (00.032319) Error (criu/cr-dump.c:2098): Dumping FAILED.
not ok 28 checkpoint and restore (bind mount, destination is symlink)
```

**Comment by alexellis** [comment]
Hi @AkihiroSuda yes we did add CONFIG_CHECKPOINT_RESTORE for @saschagrunert - then today we also added about 8x more Kernel features requested by one of his associates.

If you re-test now, it would show if the other components are enough?

CONFIG_CHECKPOINT_RESTORE
CONFIG_UNIX_DIAG
CONFIG_INET_DIAG
CONFIG_INET_UDP_DIAG
CONFIG_PACKET_DIAG
CONFIG_NETLINK_DIAG
CONFIG_NETFILTER_XT_MARK
CONFIG_INET_TCP_DIAG
CONFIG_INET_UDP_DIAG
CONFIG_INET_RAW_DIAG

Alex

**Comment by kolyshkin** [comment]
@alexellis you can run https://github.com/opencontainers/runc/blob/main/script/check-config.sh to see what's missing. On my machine (Fedora 39) it shows two missing things (both in the OPTIONAL section):

> - CONFIG_RT_GROUP_SCHED: missing

(It's not available for cgroup v2 AFAIR)

> - CONFIG_SECURITY_APPARMOR: missing

(Alternative to SELINUX, I think it's not possible to have both).

**Comment by adrianreber** [comment]
Details about the checkpoint/restore problems https://github.com/cri-o/cri-o/issues/7629#issuecomment-1951356893

**Comment by alexellis** [comment]
The Kernel may have been "born" some time ago, but isn't old, it's a relatively new patch revision.

That said, if 5.10.x and Ubuntu 22.04 are not going to work for CRIU, can you merge this as a step forward for Arm support, and keep an action item to enable CRIU e2e tests for when we have something like a 6.x Kernel available?

Would having CI for 99% of runc running on Arm CI be better, than blocking everything due to not having CRIU coverage?

**Comment by adrianreber** [comment]
Maybe it is enough to retry with the latest kernel. The latest error I see here is not related to restartable sequences and the files which are running in the test containers are using binaries from glibc 2.31 (busybox) and glibc 2.28 (debian).

At this point, with the latest kernel, it might just work.

File: update.go (hunk 14 above)
Comment (kolyshkin) [reply update.go:266]: I'm reworking this commit in #4227, hoping to come with simpler code. Still a WIP, will continue tomorrow.

Comment (kolyshkin) [reply update.go:266]: This particular part is covered with https://github.com/opencontainers/runc/pull/4226 (which I'm going to carry in one way or another in #4227).

**Comment by AkihiroSuda** [comment]
Rebased.
I guess the criu stuff can be revisited in a separate PR later.

**Comment by alexellis** [comment]
We have a 6.x Kernel on the roadmap, but it's a significant amount of labour and iteration to take a Firecracker guest Kernel and only enable the specific features for containers and ebpf, so if that resolves the issues with CRIU we will reach out when it's ready.

I'd suggest you merge this PR and skip those tests for the time being on arm64. The project's test coverage on native hardware will be in a much better position already. Would you agree?

**Comment by AkihiroSuda** [comment]
Ping @opencontainers/runc-maintainers 
Let's merge this

File: tests/integration/helpers.bash [hunk 15]
```diff
@@ -451,6 +451,22 @@ function requires() {
 				skip_me=1
 			fi
 			;;
+		cgroups_io_weight)
+			local p f f2
```
Comment (kolyshkin) [inline tests/integration/helpers.bash:457]: nit: this should be `p f1 f2`.

**Review by kolyshkin — APPROVED** [review]
LGTM overall, just one minor nit

File: tests/integration/helpers.bash (hunk 15 above)
Comment (AkihiroSuda) [reply tests/integration/helpers.bash]: fixed

File: .github/workflows/test.yml [hunk 16]
```diff
@@ -67,13 +131,12 @@ jobs:
         curl -fSsLl $REPO/Release.key | gpg --dearmor | sudo tee /etc/apt/trusted.gpg.d/devel_tools_criu.gpg > /dev/null
         echo "deb $REPO/ /" | sudo tee /etc/apt/sources.list.d/criu.list
         sudo apt update
-        sudo apt install libseccomp-dev criu sshfs
+        sudo apt -y install criu
 
-    - name: install deps (criu ${{ matrix.criu }})
-      if: matrix.criu != ''
+    - name: install CRIU (criu ${{ matrix.criu }})
+      if: ${{ matrix.os != 'actuated-arm64-6cpu-8gb' && matrix.criu == '' }}
```
Comment (lifubang) [inline .github/workflows/test.yml:137]: s/matrix.criu == ''/matrix.criu != ''/

File: .github/workflows/test.yml [hunk 17]
```diff
@@ -50,15 +50,79 @@ jobs:
...
+            os: actuated-arm64-6cpu-8gb
+          - dmz: runc_nodmz
+            os: actuated-arm64-6cpu-8gb
+
     runs-on: ${{ matrix.os }}
 
     steps:
+# https://gist.github.com/alexellis/1f33e581c75e11e161fe613c46180771#file-metering-gha-md
+# vmmeter start
+    - uses: alexellis/arkade-get@master
```
Comment (lifubang) [inline .github/workflows/test.yml:67]: Shall we add a step name here?

File: .github/workflows/test.yml [hunk 18]
```diff
@@ -50,15 +50,79 @@ jobs:
...
+      with:
+        crane: latest
+        print-summary: false
+
+    - name: Install vmmeter
+      if: matrix.os == 'actuated-arm64-6cpu-8gb'
+      run: |
+        crane export --platform linux/arm64 ghcr.io/openfaasltd/vmmeter:latest | sudo tar -xvf - -C /usr/local/bin
+
+    - uses: self-actuated/vmmeter-action@master
```
Comment (lifubang) [inline .github/workflows/test.yml:78]: Shall we add a step name here?

File: .github/workflows/test.yml (hunk 16 above)
Comment (kolyshkin) [reply .github/workflows/test.yml:137]: excellent catch, thanks!

File: .github/workflows/test.yml [hunk 19]
```diff
@@ -67,13 +131,12 @@ jobs:
         curl -fSsLl $REPO/Release.key | gpg --dearmor | sudo tee /etc/apt/trusted.gpg.d/devel_tools_criu.gpg > /dev/null
         echo "deb $REPO/ /" | sudo tee /etc/apt/sources.list.d/criu.list
         sudo apt update
```
Comment (kolyshkin) [inline .github/workflows/test.yml:133]: this `sudo apt update` needs to be removed as we now have a step above that does it

File: README.md [hunk 20]
```diff
@@ -6,6 +6,7 @@
 [![gha/validate](https://github.com/opencontainers/runc/workflows/validate/badge.svg)](https://github.com/opencontainers/runc/actions?query=workflow%3Avalidate)
 [![gha/ci](https://github.com/opencontainers/runc/workflows/ci/badge.svg)](https://github.com/opencontainers/runc/actions?query=workflow%3Aci)
 [![CirrusCI](https://api.cirrus-ci.com/github/opencontainers/runc.svg)](https://cirrus-ci.com/github/opencontainers/runc)
+<a href="https://actuated.dev"><img alt="Arm CI sponsored by Actuated" src="https://docs.actuated.dev/images/actuated-badge.png" width="120px"></img></a>
```
Comment (kolyshkin) [inline README.md:9]: nit: maybe use md syntax like in the above lines.

Comment (AkihiroSuda) [reply README.md:9]: No, the md syntax doesn't support width IIUC

File: .github/workflows/test.yml (hunk 19 above)
Comment (AkihiroSuda) [reply .github/workflows/test.yml:133]: No, the source list is updated after that

File: .github/workflows/test.yml (hunk 18 above)
Comment (AkihiroSuda) [reply .github/workflows/test.yml]: done

File: .github/workflows/test.yml (hunk 16 above)
Comment (AkihiroSuda) [reply .github/workflows/test.yml]: fixed

**Review by lifubang — APPROVED** [review]

**Comment by lifubang** [comment]
Thanks @alexellis

**Comment by alexellis** [comment]
You're welcome.
